### PR TITLE
Clustermissile recursion fix

### PR DESCRIPTION
**Pull request by SuicidalInsanity**
Fix clustermissiles spawning more clustermissiles instead of submunitions

### CODE REVIEW

File: BDArmory/Weapons/Missiles/MultiMissileLauncher.cs [hunk 1]
```diff
@@ -149,6 +150,7 @@ public void Start()
             }
             targetsAssigned = new List<TargetInfo>();
             StartCoroutine(DelayedStart());
+            if (isClusterMissile && vessel.Parts.Count == 1) isLaunchedClusterMissile = true;
```
Comment (BrettRyland) [inline BDArmory/Weapons/Missiles/MultiMissileLauncher.cs:153]: This also needs a `HighLogic.LoadedSceneIsFlight` condition as `vessel` isn't valid in the SPH.

**Review by BrettRyland — COMMENTED** [review]
I'm seeing some weird behaviour of the cluster missile with this PR.
https://www.youtube.com/watch?v=Y17ysw5DUZc
In particular:
- The launch angle as the missile detaches isn't the same as how it sits on the missile rail.
- The submunitions wobble all over the place (this might simply be an aerodynamics thing). If they didn't wobble, it's fairly likely that they'd overheat and explode with the speeds that they're reaching, so maybe they shouldn't be thrusting for that long, should thrust less or have some sort of guidance?
- For some reason, the wing of the plane with the other missile attached got destroyed when the fired missile deployed the submunitions.
- Also, I got this stacktrace when firing the missile from a standard missile rail:
```
[ERR 22:12:10.630] Input is null for field 'loadedMissileName' in config node 'MODULE'
  at System.Environment.get_StackTrace () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at KSPCommunityFixes.Performance.ConfigNodePerf.ConfigNode_AddValue2_Prefix (ConfigNode __instance, System.String name, System.String value) [0x00003] in C:\Users\helen\Source\Repos\KSPModdingLibs\KSPCommunityFixes\KSPCommunityFixes\Performance\ConfigNodePerf.cs:546 
  at ConfigNode.ConfigNode.AddValue_Patch1 (ConfigNode , System.String , System.String ) [0x00000] in <be370b73275e49439ea5e41ceef6700f>:0 
  at BaseFieldList.Save (ConfigNode node) [0x00000] in <be370b73275e49439ea5e41ceef6700f>:0 
  at PartModule.Save (ConfigNode node) [0x00000] in <be370b73275e49439ea5e41ceef6700f>:0 
  at ProtoPartModuleSnapshot.EnsureModuleValuesIsInitialized () [0x00000] in <be370b73275e49439ea5e41ceef6700f>:0 
  at ProtoPartModuleSnapshot.Save (ConfigNode node) [0x00000] in <be370b73275e49439ea5e41ceef6700f>:0 
  at ProtoPartSnapshot.Save (ConfigNode node) [0x00000] in <be370b73275e49439ea5e41ceef6700f>:0 
  at BDArmory.Weapons.Missiles.ModuleMissileRearm.PartSnapshot (Part part) [0x00045] in /home/brett/github/KSP/BDArmory/BDArmory/Weapons/Missiles/ModuleMissileRearm.cs:178 
  at BDArmory.Weapons.Missiles.ModuleMissileRearm.SpawnMissile (UnityEngine.Transform MissileTransform, System.Single offset, System.Boolean deductAmmo) [0x0008f] in /home/brett/github/KSP/BDArmory/BDArmory/Weapons/Missiles/ModuleMissileRearm.cs:54 
  at BDArmory.Weapons.Missiles.MissileLauncher+<FireReloadableMissile>d__141.MoveNext () [0x00070] in /home/brett/github/KSP/BDArmory/BDArmory/Weapons/Missiles/MissileLauncher.cs:1100 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <2425394cf09642369e2109953e31f62b>:0 
  at UnityEngine.MonoBehaviour.StartCoroutineManaged2 (System.Collections.IEnumerator enumerator) [0x00000] in <2425394cf09642369e2109953e31f62b>:0 
  at UnityEngine.MonoBehaviour.StartCoroutine (System.Collections.IEnumerator routine) [0x00000] in <2425394cf09642369e2109953e31f62b>:0 
  at BDArmory.Weapons.Missiles.MissileLauncher.FireMissile () [0x00188] in /home/brett/github/KSP/BDArmory/BDArmory/Weapons/Missiles/MissileLauncher.cs:1079 
  at BDArmory.Control.MissileFire.FireCurrentMissile (BDArmory.Weapons.Missiles.MissileBase missile, System.Boolean checkClearance) [0x0028c] in /home/brett/github/KSP/BDArmory/BDArmory/Control/MissileFire.cs:3139 
  at BDArmory.Control.MissileFire.FireMissile () [0x00076] in /home/brett/github/KSP/BDArmory/BDArmory/Control/MissileFire.cs:3208 
  at BDArmory.Control.MissileFire.FireMissileManually (System.Boolean mainTrigger) [0x00070] in /home/brett/github/KSP/BDArmory/BDArmory/Control/MissileFire.cs:3230 
  at BDArmory.Control.MissileFire.OnUpdate () [0x00167] in /home/brett/github/KSP/BDArmory/BDArmory/Control/MissileFire.cs:1403 
  at Part.ModulesOnUpdate () [0x00000] in <be370b73275e49439ea5e41ceef6700f>:0 
  at Part.Update () [0x00000] in <be370b73275e49439ea5e41ceef6700f>:0 
  UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
  ModuleManager.UnityLogHandle.InterceptLogHandler:LogFormat(LogType, Object, String, Object[])
  UnityEngine.Debug:LogError(Object)
  KSPCommunityFixes.Performance.ConfigNodePerf:ConfigNode_AddValue2_Prefix(ConfigNode, String, String) (at C:/Users/helen/Source/Repos/KSPModdingLibs/KSPCommunityFixes/KSPCommunityFixes/Performance/ConfigNodePerf.cs:546)
  ConfigNode:ConfigNode.AddValue_Patch1(ConfigNode, String, String)
  BaseFieldList:Save(ConfigNode)
  PartModule:Save(ConfigNode)
  ProtoPartModuleSnapshot:EnsureModuleValuesIsInitialized()
  ProtoPartModuleSnapshot:Save(ConfigNode)
  ProtoPartSnapshot:Save(ConfigNode)
  BDArmory.Weapons.Missiles.ModuleMissileRearm:PartSnapshot(Part) (at /home/brett/github/KSP/BDArmory/BDArmory/Weapons/Missiles/ModuleMissileRearm.cs:178)
  BDArmory.Weapons.Missiles.ModuleMissileRearm:SpawnMissile(Transform, Single, Boolean) (at /home/brett/github/KSP/BDArmory/BDArmory/Weapons/Missiles/ModuleMissileRearm.cs:54)
  BDArmory.Weapons.Missiles.<FireReloadableMissile>d__141:MoveNext() (at /home/brett/github/KSP/BDArmory/BDArmory/Weapons/Missiles/MissileLauncher.cs:1100)
  UnityEngine.MonoBehaviour:StartCoroutine(IEnumerator)
  BDArmory.Weapons.Missiles.MissileLauncher:FireMissile() (at /home/brett/github/KSP/BDArmory/BDArmory/Weapons/Missiles/MissileLauncher.cs:1079)
  BDArmory.Control.MissileFire:FireCurrentMissile(MissileBase, Boolean) (at /home/brett/github/KSP/BDArmory/BDArmory/Control/MissileFire.cs:3139)
  BDArmory.Control.MissileFire:FireMissile() (at /home/brett/github/KSP/BDArmory/BDArmory/Control/MissileFire.cs:3208)
  BDArmory.Control.MissileFire:FireMissileManually(Boolean) (at /home/brett/github/KSP/BDArmory/BDArmory/Control/MissileFire.cs:3230)
  BDArmory.Control.MissileFire:OnUpdate() (at /home/brett/github/KSP/BDArmory/BDArmory/Control/MissileFire.cs:1403)
  Part:ModulesOnUpdate()
  Part:Update()
```
from `loadedMissileName` being empty in the MultiMissileLauncher module in the craft file. Do you have an updated BDA-E that isn't released yet?

**Comment by SuicidalInsanity** [comment]
YT vid is set to private, I can't watch it.
Submunition wobble and overvelocity is a submunition .cfg config issue, adjusting .cfg values fixed the aberrant behavior.
I haven't seen the off-angle launch on my end, so not sure what's causing that.
I think I've ID'd what's causing the 'other wing get destroyed when the missile deploys' issue, but haven't been able to replicate, so uncertain if the fix for that is working.

**Comment by BrettRyland** [comment]
> YT vid is set to private, I can't watch it.

Ah, I thought private meant that you could watch it if you had the link. I've switched it to unlisted, so you should be able to see it now.

> I think I've ID'd what's causing the 'other wing get destroyed when the missile deploys' issue, but haven't been able to replicate, so uncertain if the fix for that is working.

Further testing was somewhat inconsistent whether the wings broke or not and which wing(s) broke if they did. So I'm not sure what's going on there, but that plane doesn't lose its wings when firing other missiles, so I think it's specific to the cluster missile.

File: BDArmory/Weapons/Missiles/MissileLauncher.cs [hunk 2]
```diff
@@ -1064,13 +1064,44 @@ public override void FireMissile()
...
+                        TimeFired = Time.time;
+                        part.decouple(0);
+                        part.Unpack();
+                        TargetPosition = vessel.ReferenceTransform.position + vessel.ReferenceTransform.up * 5000; //set initial target position so if no target update, missileBase will count a miss if it nears this point or is flying post-thrust
+                        MissileLaunch();
+                        BDATargetManager.FiredMissiles.Add(this);
+                        if (wpm != null)
+                        {
+                            wpm.heatTarget = TargetSignatureData.noTarget;
+                            GpsUpdateMax = wpm.GpsUpdateMax;
```
[the remark this answers is not in the document]
Comment (SuicidalInsanity) [reply BDArmory/Weapons/Missiles/MissileLauncher.cs]: oops, yeah, that isn't meant for this PR.

**Pull request merged**